ledger/acctupdates_test.go

```diff
@@ -2641,3 +2639,91 @@ func TestAcctUpdatesLookupLatestCacheRetry(t *testing.T) {
+	}
+	proto := config.Consensus[protocol.ConsensusCurrentVersion]
+
+	accts := []map[basics.Address]basics.AccountData{ledgertesting.RandomAccounts(20, true)}
+
+	pooldata := basics.AccountData{}
+	pooldata.MicroAlgos.Raw = 1000 * 1000 * 1000 * 1000
+	pooldata.Status = basics.NotParticipating
+	accts[0][testPoolAddr] = pooldata
+
+	sinkdata := basics.AccountData{}
+	sinkdata.MicroAlgos.Raw = 1000 * 1000 * 1000 * 1000
+	sinkdata.Status = basics.NotParticipating
+	accts[0][testSinkAddr] = sinkdata
+
+	ml := makeMockLedgerForTracker(t, true, 10, protocol.ConsensusCurrentVersion, accts)
+	defer ml.Close()
+
+	conf := config.GetDefaultLocal()
+	au, ao := newAcctUpdates(t, ml, conf, ".")
+	defer au.close()
+	defer ao.close()
+
+	// cover 10 genesis blocks
+	rewardLevel := uint64(0)
+	for i := 1; i < 10; i++ {
+		accts = append(accts, accts[0])
+	}
+
+	// lastCreatableID stores asset or app max used index to get rid of conflicts
+	lastCreatableID := crypto.RandUint64() % 512
+	knownCreatables := make(map[basics.CreatableIndex]bool)
+
+	start := basics.Round(10)
+	end := basics.Round(2*proto.MaxBalLookback + 15)
+	for i := start; i < end; i++ {
+		rewardLevelDelta := crypto.RandUint64() % 3
+		rewardLevel += rewardLevelDelta
+		var updates ledgercore.AccountDeltas
+		var totals map[basics.Address]ledgercore.AccountData
+		base := accts[i-1]
+		updates, totals, lastCreatableID = ledgertesting.RandomDeltasBalancedFull(1, base, rewardLevel, lastCreatableID)
+		prevRound, prevTotals, err := au.LatestTotals()
+		require.Equal(t, i-1, prevRound)
+		require.NoError(t, err)
+
+		newPool := totals[testPoolAddr]
+		newPool.MicroAlgos.Raw -= prevTotals.RewardUnits() * rewardLevelDelta
+		updates.Upsert(testPoolAddr, newPool)
+		totals[testPoolAddr] = newPool
+		newAccts := applyPartialDeltas(base, updates)
+
+		blk := bookkeeping.Block{
+			BlockHeader: bookkeeping.BlockHeader{
+				Round: basics.Round(i),
+			},
+		}
+		blk.RewardsLevel = rewardLevel
+		blk.CurrentProtocol = protocol.ConsensusCurrentVersion
+
+		delta := ledgercore.MakeStateDelta(&blk.BlockHeader, 0, updates.Len(), 0)
+		delta.Accts.MergeAccounts(updates)
+		delta.Creatables = creatablesFromUpdates(base, updates, knownCreatables)
+
+		delta.Totals = accumulateTotals(t, protocol.ConsensusCurrentVersion, []map[basics.Address]ledgercore.AccountData{totals}, rewardLevel)
+		ml.trackers.newBlock(blk, delta)
+		accts = append(accts, newAccts)
+	}
+
+	ml.trackers.lastFlushTime = time.Time{}
+
+	ml.trackers.committedUpTo(2*basics.Round(proto.MaxBalLookback) + 14)
+	ml.trackers.waitAccountsWriting()
+
+	var dbOnlineRoundParams []ledgercore.OnlineRoundParamsData
+	err := ao.dbs.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		dbOnlineRoundParams, err = accountsOnlineRoundParams(tx)
+		return err
+	})
+	require.NoError(t, err)
+	require.Equal(t, dbOnlineRoundParams, ao.onlineRoundParamsData[:basics.Round(proto.MaxBalLookback)])
```

Comment:
this is good, but not enough.
please add the following:
1. Add protocol change while adding new blocks
2. Save totals and protocols while adding blocks
3. Compare saved values against actual values by using ao.OnlineTotals(rnd) for Latest-proto.MaxBalLookback... Latest
4.Calculate expected last persisted round. Check it and below are not in the DB.

Reply:
what should be the function signature of ao.OnlineTotals?

4. is already covered.

Reply:
```
// OnlineTotals returns the online totals of all accounts at the end of round rnd.
func (l *Ledger) OnlineTotals(rnd basics.Round) (basics.MicroAlgos, error) {
	l.trackerMu.RLock()
	defer l.trackerMu.RUnlock()
	return l.accts.OnlineTotals(rnd)
}
```
should use online accounts not acct updates.

> is already covered
No, the test checks the cache against the db. But it does not check on how many entries it should have (MaxLookback + small delta that needs to be calculated).

Reply:
No, it checks that after a db write there are exactly 320 entries in the db and that they are being stored by the cache.

Reply:
oh, right, I misread. great!